### PR TITLE
Experiment/ppo vispendulum self boost

### PR DESCRIPTION
**Pull request by antonbolychev**
# Self-Boosting PPO Agent Enhancement

## Hypothesis
Passing previous checkpoints as fallback policies to CALFWrapper doesn't degrade agent perfomance

## Enviroment

- Visual Pendulum

### New Directories
- `run/ppo_vis_pendulum_self_boost/`
  - Training and evaluation scripts
  - Configuration files
  - Results and artifact storage

- `analysis/ppo_vis_pendulum_self_boost/`
  - Performance analysis notebooks
  - Comparative evaluations
  - Visualization tools

### Core Changes
- `src/`
  - Integration of fallback policy mechanism
  - Core implementation modifications

### Documentation
- [ ] `run/ppo_vis_pendulum_self_boost/` README includes:
  - Experimental hypothesis
  - Implementation details
  - Artifact management strategy
  - Execution instructions

### Analysis
- [ ] `analysis/ppo_vis_pendulum_self_boost/` contains:
  - Functional analysis notebooks
  - Clear methodology documentation
  - Comprehensive result interpretation

### Source Code
- [ ] Changes in `src/` are:
  - Properly documented
  - Scope clearly defined
  - Integration points specified

### CODE REVIEW

File: run/ppo_vispendulum_self_boost/README.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
+# Visual PPO with CALFWrapper evaluation
+
+> [!IMPORTANT]  
+> It's supposed that this subfolder is your working directory: 
+> do not forget to apply in your terminal `cd run/ppo_pendulum_calf_wrapper_eval`
+
```
Comment (huyhoang251296) [inline run/ppo_vispendulum_self_boost/README.md:6]: Test comment

File: run/ppo_vispendulum_self_boost/pendulum_visual_ppo.py
Comment (antonbolychev) [inline run/ppo_vispendulum_self_boost/pendulum_visual_ppo.py:1]: One more comment about the argument parser

I reccoment you to use another approach that I find in CleanRL
For instance see
https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ddpg_continuous_action.py

They are using tyro instead of argument parser

There is no need to follow the approach at THE moment and you can resolve the conversation not to waste more time for refactoring, but tyro can save you loads of time =)

https://brentyi.github.io/tyro/

Another option here is https://typer.tiangolo.com/

File: src/wrapper/calf_wrapper.py [hunk 2]
```diff
@@ -130,3 +174,81 @@ def reset(self, **kwargs):
...
+    def reset_internal_params(self):
+        if self.relax_prob_episode_activated:
+            self.relax_prob_step_factor = self.relax_prob_base_step_factor
+            self.initial_relax_prob = np.clip(
+                self.initial_relax_prob + self.initial_relax_prob * self.relax_prob_episode_factor,
+                0, 1)
+            
+            # Reset Relax decay mechanism
+            self.relax_decay.reset(self.initial_relax_prob)
+            self.relax_prob = self.relax_decay.get_relax_prob()
```
[the remark this answers is not in the document]
Comment (huyhoang251296) [reply src/wrapper/calf_wrapper.py:230]: At this moment, the RelaxProb only supports the decay of relax_prob over time_step. It is not aware of the change to a new episode. The logic at 211 is still needed now, although we never reach that condition when running the evaluation.

[the remark this answers is not in the document]
Comment (antonbolychev) [reply src/wrapper/calf_wrapper.py:229]: Why can't you just send all the parameters like  self.initial_relax_prob self.relax_prob_episode_factor etc and do this in reset function of RelaxProb class?

[the remark this answers is not in the document]
Comment (antonbolychev) [reply src/wrapper/calf_wrapper.py:229]: image

This guys can be inside of RelaxProb.reset without any backfires

File: run/ppo_vispendulum_self_boost/ppo_vispendulum.py [hunk 3]
```diff
@@ -23,51 +27,56 @@
...
 # Define the hyperparameters for PPO
 ppo_hyperparams = {
     "learning_rate": 4e-4,  # The step size used to update the policy network. Lower values can make learning more stable.
     "n_steps": n_steps,  # Number of steps to collect before performing a policy update. Larger values may lead to more stable updates.
     "batch_size": 512,  # Number of samples used in each update. Smaller values can lead to higher variance, while larger values stabilize learning.
-    "gamma": 0.99,  # Discount factor for future rewards. Closer to 1 means the agent places more emphasis on long-term rewards.
+    "gamma": 0.98,  # Discount factor for future rewards. Closer to 1 means the agent places more emphasis on long-term rewards.
     "gae_lambda": 0.9,  # Generalized Advantage Estimation (GAE) parameter. Balances bias vs. variance; lower values favor bias.
-    "clip_range": 0.2,  # Clipping range for the PPO objective to prevent large policy updates. Keeps updates more conservative.
+    "clip_range": 0.01,  # Clipping range for the PPO objective to prevent large policy updates. Keeps updates more conservative.
```
Comment (antonbolychev) [inline run/ppo_vispendulum_self_boost/ppo_vispendulum.py:53]: Again WHY 0.01?

File: src/wrapper/calf_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (huyhoang251296) [reply src/wrapper/calf_wrapper.py:229]: Yes, you are right. But to do that I have to modify so many things in the ppo_pendulum_calf_wrapper_eval which I am trying to keep far away from it. I can update on Monday 16.12

File: run/ppo_vispendulum_self_boost/ppo_vispendulum.py (hunk 3 above)
Comment (huyhoang251296) [reply run/ppo_vispendulum_self_boost/ppo_vispendulum.py:53]: After randomizing initial state of the Pendulum, Visual Pendulum PPO couldn't learn well with the default config.
I reduced clip_range and change other configs to make it worked.
Since it worked properly, I haven't checked the training again and started refactoring everything...

File: run/ppo_vispendulum_self_boost/pendulum_visual_ppo.py [hunk 4]
```diff
@@ -148,7 +164,7 @@ def _init():
         # If --console flag is set, disable the plot and just save the data
         if args.console:
             plotting_callback.figure = None  # Disable plotting
-            print("Console mode: Graphical output disabled. Episode rewards will be saved to 'logs/episode_rewards.csv'.")
+            print("Console mode: Graphical output disabled. Episode rewards will be saved to 'episode_rewards.csv'.")
```
[the remark this answers is not in the document]
Comment (huyhoang251296) [reply run/ppo_vispendulum_self_boost/pendulum_visual_ppo.py]: Yes we have, it should be `logs/episode_rewards.csv`

File: run/ppo_vispendulum_self_boost/pendulum_visual_ppo.py [hunk 5]
```diff
@@ -190,21 +213,32 @@ def _init():
...
+        "state": [],
+        "action": [],
+        "reward": [],
+        "accumulated_reward": [],
+    }
+    accumulated_reward = 0
 
     # Run the simulation with the trained agent
-    for _ in range(3000):
+    for _ in range(1000):
```
[the remark this answers is not in the document]
Comment (huyhoang251296) [reply run/ppo_vispendulum_self_boost/pendulum_visual_ppo.py:241]: 3000 steps take a lot of time while 1000 steps can help us have enough information and save 3 times running time.
From that point, I would like to make a new pattern for all of my experiments further

File: pyproject.toml [hunk 6]
```diff
@@ -9,58 +9,20 @@ packages = [
...
-nvidia-cudnn-cu12 = "9.1.0.70"
-nvidia-cufft-cu12 = "11.2.1.3"
-nvidia-curand-cu12 = "10.3.5.147"
-nvidia-cusolver-cu12 = "11.6.1.9"
-nvidia-cusparse-cu12 = "12.3.1.170"
-nvidia-nccl-cu12 = "2.21.5"
-nvidia-nvjitlink-cu12 = "12.4.127"
-nvidia-nvtx-cu12 = "12.4.127"
-opencv-python = "4.10.0.84"
-opencv-python-headless = "4.10.0.84"
```
Comment (antonbolychev) [inline pyproject.toml:45]: Please, return back the package

**Comment by antonbolychev** [comment]
also fix this thing too


image

File: run/ppo_pendulum_calf_wrapper_eval/README.md
Comment (antonbolychev) [inline run/ppo_pendulum_calf_wrapper_eval/README.md:1]: The README should be rewritten!

Pavel says absolutely impossible to follow

File: run/ppo_vispendulum_self_boost/ppo_vispendulum.py (hunk 3 above)
Comment (huyhoang251296) [reply run/ppo_vispendulum_self_boost/ppo_vispendulum.py:53]: To be specific, the default pendulum_ppo_visual does not work without `--normalize` option. In that case, the `--normalize` should always be True in the script.

However, In my configuration, PPO can learn with or without `--normalize` option but it takes longer for rewards to converge to 0

-> I will convert to the original config and  `--normalize` is True as default.

**Comment by huyhoang251296** [comment]
Feedback
image

**Comment by huyhoang251296** [comment]
Feedback
image

**Comment by huyhoang251296** [comment]
image

**Comment by huyhoang251296** [comment]
image
